### PR TITLE
Trang Frego - JS Adagrams

### PR DESCRIPTION
**Pull request by tfrego**
# JS Adagrams
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
--- | ---
What patterns were you able to use from your Ruby knowledge to apply to JavaScript? | iterating through arrays
Did you need to use different strategies to find information online about JavaScript than you would for Ruby? |  I used similar strategies which consisted of reading through the JavaScript documentation on MDN and google searches on StackOverflow
What was a challenge you were able to overcome on this assignment? | Becoming comfortable with the new syntax
What is your favorite thing about learning a new programming language? | It makes me use more logic thinking on how I want to obtain my solution instead of using a method that was easily provided in other languages
What is your least favorite thing about learning a new programming language? |  learning new syntax and being very slow at knowing what I can or cannot do 
Do you have any recommendations on how we could improve this project for the next cohort? | Provide debugging tools =)

### CODE REVIEW

File: src/adagrams.js [hunk 1]
```diff
@@ -1,8 +1,141 @@
+// count elements in array
+const count = (arr, element) => {
```
Comment (Hamled) [inline src/adagrams.js:2]: Also could be done with:

```js
Array.prototype.count = function(element) {
  return this.filter(e => e === element).length;
};
```

and then using `arr.count(element)` to call it.

Of course, while this is shorter code it does require creating a new array (the result of the `filter` function), so your approach would be preferable for large arrays or very frequent use (e.g. thousands of times per second).

File: src/adagrams.js [hunk 2]
```diff
@@ -1,8 +1,141 @@
...
+      } else {
+        return true;
+      }
+    });
+
+    let answer = true;
+
+    letterResults.forEach(function (item) {
+      if (item === false) {
+        answer = false;
```
Comment (Hamled) [inline src/adagrams.js:42]: I would suggest looking into the [`Array.prototype.every`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every) function, as it could shorten this function significantly.

Additionally, what is the Big-O time complexity of this function (keep in mind the code in your `count` function)?

**Comment by Hamled** [comment]
# JS Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | :white_check_mark:<br>I agree we should have good debugging tools available at the beginning of our JS curriculum.
Small commits with meaningful commit messages | :white_check_mark:
**Code Requirements** |
`drawLetters` method | :white_check_mark:
Uses appropriate data structure to store the letter distribution | :white_check_mark:
All tests for `drawLetters` pass | :white_check_mark:
`usesAvailableLetters` method | :white_check_mark:
All tests for `usesAvailableLetters` pass | :white_check_mark:
`scoreWord` method | :white_check_mark:
Uses appropriate data structure to store the letter scores | :white_check_mark:
All tests for `scoreWord` pass | :white_check_mark:
`highestScoreFrom` method | :white_check_mark:
Appropriately handles edge cases for tie-breaking logic | :white_check_mark:
All tests for `highestScoreFrom` pass | :white_check_mark:
**Overall** | Looks good! I like the approach that you took for `highestScoreFrom`, I had not thought of doing it that way before.

File: src/adagrams.js [hunk 3]
```diff
@@ -1,8 +1,141 @@
...
+    }
+  });
+  return sum;
+};
+
 const Adagrams = {
   drawLetters() {
-    // Implement this method for wave 1
+    let drawnLetters = [];
+    let letterPool = 'AAAAAAAAABBCCDDDDEEEEEEEEEEEEFFGGGHHIIIIIIIIIJKLLLLMMNNNNNNOOOOOOOOPPQRRRRRRSSSSTTTTTTUUUUVVWWXYYZ'.split('')
```
Comment (Hamled) [inline src/adagrams.js:15]: While it might not seem like it would be appropriate, we can actually declare both of these variables with `const`.

The reason for this is because `const` only restricts re-assigning the variable itself to reference another value -- it doesn't restrict modifying the existing value. For this reason the calls to `push` and `slice` later in this function are fine even with `const`.

In fact, all of the variables in this function could be declared with `const`, except for `i` because it is modified by `i += 1` which is a form of re-assignment.
